### PR TITLE
Added manage legacy volume & snapshot feature

### PR DESCRIPTION
**Pull request by sonivi**
Added 'importVol=<3par_volume|3par_snapshot>' driver option.
This can be used for both import legacy volume as well as snapshot

### CODE REVIEW

File: hpedockerplugin/hpe_storage_api.py [hunk 1]
```diff
@@ -159,6 +159,26 @@ def volumedriver_create(self, name, opts=None):
                     LOG.error(msg)
                     return json.dumps({u"Err": six.text_type(msg)})
 
+            # mutually exclusive options check
+            mutually_exclusive_list = ['virtualCopyOf', 'cloneOf', 'qos-name']
+            input_list = contents['Opts'].keys()
```
Comment (nilangekarss) [inline hpedockerplugin/hpe_storage_api.py]: @sonivi This will fail in python3, Satish's code is already merged that is what i can see

Comment (nilangekarss) [reply hpedockerplugin/hpe_storage_api.py]: @sonivi can you also modify create_help.txt to accommodate information on import. Rest changes looks fine for me

Comment (sonivi) [reply hpedockerplugin/hpe_storage_api.py]: comment addressed
Also addded unit test case

**Comment by nilangekarss** [comment]
@sonivi Hi Vivek,
Please add a check to see, if the snapshot to be imported is read write or read only snapshot.
If snapshot is RO then throw an error saying only import of RW snapshot is supported. This is because, if we try to mount the RO snapshot, mount request will fail. In other words, in case of K8S environment POD creation will not be successful. Hence no point in importing snapshot.

@wdurairaj What is your opinion William?

**Comment by wdurairaj** [comment]
@sonivi , @nilangekarss 

I wanted the RW/RO snapshot(s) to be managed as it is , and further allowed for mount during a mount step. We will not add any specific validation to prevent mount for RO snapshot, because the containerized application which will mount will atleast get a chance to read the data.

File: test/createvolume_tester.py [hunk 2]
```diff
@@ -41,6 +41,45 @@ def setup_mock_objects(self):
...
+
+
+class TestImportVolumeOtherOption(CreateVolumeUnitTest):
+    def check_response(self, resp):
+        self._test_case.assertNotEqual(resp, {u"Err": ''})
+
+    def get_request_params(self):
+        return {"Name": "abc_vol",
+                "Opts": {"importVol": "vvk_vol",
+                         "size": "2"}}
```
Comment (wdurairaj) [inline test/createvolume_tester.py]: @sonivi 
Are we allowing the -o importVol=<3parvol|snapshot> and -o size=X
meaning after importing of a snapshot/volume we are changing the size of the imported vol/snap ?

**Comment by wdurairaj** [comment]
Questions
1. if the volume/snapshot is in use (active VLUN's ) present, do we still allow that volume/snapshot to be managed by the plugin?

2. If the snapshot/volume import fails, what will the state of the original snapshot/volume, will the volume name be renamed ?

File: test/createvolume_tester.py (hunk 2 above)
Comment (wdurairaj) [reply test/createvolume_tester.py]: @sonivi ,

Thanks for the clarification. I did'nt notice the assertNotEquals() .. sorry about that.

**Comment by sonivi** [comment]
ans1) yes importing in-use volume / snapshot will fail import operation
ans2) any failure in managing will reverted the volume/snapshot back to its original state

Taken care here:
[](https://github.com/sonivi/python-hpedockerplugin/blob/plugin_v2/hpedockerplugin/volume_manager.py#L339)

**Comment by sonivi** [comment]
[](https://github.com/sonivi/python-hpedockerplugin/blob/plugin_v2/hpedockerplugin/volume_manager.py#L339)

**Comment by sonivi** [comment]
https://github.com/sonivi/python-hpedockerplugin/blob/plugin_v2/hpedockerplugin/volume_manager.py#L339

**Pull request merged**